lib/hammer_cli_foreman/api/basic_auth.rb

```diff
@@ -0,0 +1,68 @@
+module HammerCLIForeman
+  module Api
+    module BasicAuth
+      def authenticate(request, args)
+        if HammerCLI.interactive?
+          get_user
+          get_password
+        end
+        super
+      end
+
+      def error(ex)
+        return unless ex.is_a?(RestClient::Unauthorized)
+
+        clear
+        message = _('Invalid username or password.')
+        begin
+          response_msg = JSON.parse(ex.response.body)['error']
+        rescue
+        end
+        UnauthorizedError.new(response_msg&.dig('message') || message)
```

Comment:
```shell
# hammer -d auth login basic-external -u user -p badpass
----B<-----SNIP----B<-----
[DEBUG 2023-02-10T11:12:49 API] Using authenticator: HammerCLIForeman::Api::SessionAuthenticatorWrapper
[ERROR 2023-02-10T11:12:49 API] 401 Unauthorized
[DEBUG 2023-02-10T11:12:49 API] {
    "error" => "External authentication did not pass."
}
[DEBUG 2023-02-10T11:12:49 Exception] Using exception handler HammerCLIForeman::ExceptionHandler#handle_general_exception
[ERROR 2023-02-10T11:12:49 Exception] Error: undefined method `dig' for "External authentication did not pass.":String
Error: undefined method `dig' for "External authentication did not pass.":String
[ERROR 2023-02-10T11:12:49 Exception]

NoMethodError (undefined method `dig' for "External authentication did not pass.":String):
    /usr/share/gems/gems/hammer_cli_foreman-3.6.0.pre.develop/lib/hammer_cli_foreman/api/basic_auth.rb:21:in `error'
    /usr/share/gems/gems/hammer_cli_foreman-3.6.0.pre.develop/lib/hammer_cli_foreman/api/session_authenticator_wrapper.rb:66:in `error'
    /usr/share/gems/gems/apipie-bindings-0.5.0/lib/apipie_bindings/api.rb:248:in `rescue in http_call'
    /usr/share/gems/gems/apipie-bindings-0.5.0/lib/apipie_bindings/api.rb:238:in `http_call'
    /usr/share/gems/gems/apipie-bindings-0.5.0/lib/apipie_bindings/api.rb:190:in `call_action'
    /usr/share/gems/gems/apipie-bindings-0.5.0/lib/apipie_bindings/api.rb:185:in `call'
    /usr/share/gems/gems/apipie-bindings-0.5.0/lib/apipie_bindings/action.rb:14:in `call'
    /usr/share/gems/gems/hammer_cli_foreman-3.6.0.pre.develop/lib/hammer_cli_foreman/api/connection.rb:36:in `login'
    /usr/share/gems/gems/hammer_cli_foreman-3.6.0.pre.develop/lib/hammer_cli_foreman/authenticate/login.rb:12:in `execute_with_params'
    /usr/share/gems/gems/hammer_cli_foreman-3.6.0.pre.develop/lib/hammer_cli_foreman/auth.rb:41:in `execute'
    /usr/share/gems/gems/clamp-1.1.2/lib/clamp/command.rb:63:in `run'
    /usr/share/gems/gems/hammer_cli-3.6.0.pre.develop/lib/hammer_cli/abstract.rb:103:in `run'
    /usr/share/gems/gems/clamp-1.1.2/lib/clamp/subcommand/execution.rb:11:in `execute'
    /usr/share/gems/gems/clamp-1.1.2/lib/clamp/command.rb:63:in `run'
    /usr/share/gems/gems/hammer_cli-3.6.0.pre.develop/lib/hammer_cli/abstract.rb:103:in `run'
    /usr/share/gems/gems/clamp-1.1.2/lib/clamp/subcommand/execution.rb:11:in `execute'
    /usr/share/gems/gems/clamp-1.1.2/lib/clamp/command.rb:63:in `run'
    /usr/share/gems/gems/hammer_cli-3.6.0.pre.develop/lib/hammer_cli/abstract.rb:103:in `run'
    /usr/share/gems/gems/clamp-1.1.2/lib/clamp/subcommand/execution.rb:11:in `execute'
    /usr/share/gems/gems/clamp-1.1.2/lib/clamp/command.rb:63:in `run'
    /usr/share/gems/gems/hammer_cli-3.6.0.pre.develop/lib/hammer_cli/abstract.rb:103:in `run'
    /usr/share/gems/gems/clamp-1.1.2/lib/clamp/command.rb:132:in `run'
    /usr/share/gems/gems/hammer_cli-3.6.0.pre.develop/bin/hammer:147:in `<top (required)>'
    /usr/bin/hammer:23:in `load'
    /usr/bin/hammer:23:in `<main>'
```
```suggestion
        message = _('Invalid username or password.')
        begin
          response_msg = JSON.parse(ex.response.body)
        rescue
        end
        UnauthorizedError.new(response_msg&.dig('error') || message)
```
```

Reply:
Right :/ I'm going to rewrite it a bit. The problem is that we don't have consistent response here :/ `JSON.parse(ex.response.body)['error']` can be either string or a hash with `message` entry.

Reply:
I've changed the message. It now should show the response from the server (if any) plus the default message about invalid credentials (which is most obvious reason to fail for basic auth).